tests/modules/moduleconfigs.c

```diff
@@ -0,0 +1,136 @@
+    REDISMODULE_NOT_USED(privdata);
+    return strval;
+}
+int setStringConfigCommand(const char *name, RedisModuleString *new, void *privdata, const char **err) {
+    REDISMODULE_NOT_USED(name);
+    REDISMODULE_NOT_USED(err);
+    REDISMODULE_NOT_USED(privdata);
+    size_t len;
+    /* Address sanitizer will complain about a leak if we're not properly freeing core side. */
+    if (!strcasecmp(RedisModule_StringPtrLen(new, &len), "rejectisfreed")) {
+        return REDISMODULE_ERR;
+    }
+    if (strval) RedisModule_Free(strval);
+    RedisModule_RetainString(NULL, new);
+    strval = new;
+    return REDISMODULE_OK;
+}
+
+int getEnumConfigCommand(const char *name, void *privdata) {
+    REDISMODULE_NOT_USED(name);
+    REDISMODULE_NOT_USED(privdata);
+    return enumval;
+}
+
+int setEnumConfigCommand(const char *name, int val, void *privdata, const char **err) {
+    REDISMODULE_NOT_USED(name);
+    REDISMODULE_NOT_USED(err);
+    REDISMODULE_NOT_USED(privdata);
+    /* We don't have to do any verification here, the core makes sure its in the range of enum_vals we provide */
+    enumval = val;
+    return REDISMODULE_OK;
+}
+
+int boolApplyFunc(RedisModuleCtx *ctx, void *privdata, const char **err) {
+    REDISMODULE_NOT_USED(ctx);
+    REDISMODULE_NOT_USED(privdata);
+    if (mutable_bool_val && immutable_bool_val) {
+        *err = "Bool configs cannot both be yes.";
+        return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+int longlongApplyFunc(RedisModuleCtx *ctx, void *privdata, const char **err) {
+    REDISMODULE_NOT_USED(ctx);
+    REDISMODULE_NOT_USED(privdata);
+    if (longval == memval) {
+        *err = "These configs cannot equal each other.";
+        return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+int RedisModule_OnLoad(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    REDISMODULE_NOT_USED(argv);
+    REDISMODULE_NOT_USED(argc);
+
+    if (RedisModule_Init(ctx, "moduleconfigs", 1, REDISMODULE_APIVER_1) == REDISMODULE_ERR) return REDISMODULE_ERR;
+
+    if (RedisModule_RegisterBoolConfig(ctx, "mutable_bool", 1, REDISMODULE_CONFIG_DEFAULT, getBoolConfigCommand, setBoolConfigCommand, boolApplyFunc, &mutable_bool_val) == REDISMODULE_ERR) {
+        return REDISMODULE_ERR;
+    }
+    /* Immutable config here. */
+    if (RedisModule_RegisterBoolConfig(ctx, "immutable_bool", 0, REDISMODULE_CONFIG_IMMUTABLE, getBoolConfigCommand, setBoolConfigCommand, boolApplyFunc, &immutable_bool_val) == REDISMODULE_ERR) {
+        return REDISMODULE_ERR;
+    }
+    RedisModuleString *default_string = RedisModule_CreateString(ctx, "\x73\x65\x63\x72\x65\x74\x20\x70\x61\x73\x73\x77\x6f\x72\x64", 15);
+    if (RedisModule_RegisterStringConfig(ctx, "string", default_string, REDISMODULE_CONFIG_DEFAULT, getStringConfigCommand, setStringConfigCommand, NULL, NULL) == REDISMODULE_ERR) {
+        return REDISMODULE_ERR;
+    }
+    RedisModule_FreeString(ctx, default_string);
```

Comment:
in theory, it would been much easier if the default value for strings would have been a plain `char*`, so modules would have been able to use a static string and avoid all the mess around heap allocations.
however, doing that would mean that default values can't contain NULL chars. maybe that's acceptable?

Reply:
I think the current implementation makes more sense of passing in a String. I would somewhat prefer that we call decrRefcount though within `RedisModule_RegisterStringConfig` so that you don't have to worry about freeing it otherwise.

Reply:
I think the current implementation makes more sense of passing in a String. I would somewhat prefer that we call decrRefcount though within RedisModule_RegisterStringConfig so that you don't have to worry about freeing it otherwise.

Reply:
i understand that for the majority of the cases that would be easier to use, but the interface seems odd to me.
@yossigo @MeirShpilraien WDYT?

Reply:
@oranagra I think that for configuration it makes sense to pass `const char*` and its acceptable that it can not contains NULL chars.

Reply:
Agree with @MeirShpilraien, I don't see why a `const char *` wouldn't be enough.
But if we do have a good reason to use a RedisModuleString, I agree with @madolson we should transfer ownership so the caller doesn't need to free it.

Reply:
i think both options are a bit odd, but i suppose it's unlikely that a module will want a default value for a config to contain null chars, and i suppose ones that do, can find an ugly way around it.
so maybe a plain `char*` would be most convenient.